pydantic/schema.py

```diff
@@ -405,16 +405,23 @@ def field_type_schema(
         definitions.update(f_definitions)
         return {'type': 'array', 'uniqueItems': True, 'items': f_schema}, definitions
     elif field.shape is Shape.MAPPING:
+        dict_schema = {'type': 'object'}
+        key_pattern = None
+        regex = getattr(field.key_field.type_, 'regex', None)
+        if regex:
+            key_pattern = regex.pattern
         f_schema, f_definitions = field_singleton_schema(
             field, by_alias=by_alias, model_name_map=model_name_map, ref_prefix=ref_prefix
         )
         definitions.update(f_definitions)
-        if f_schema:
-            # The dict values are not simply Any
-            return {'type': 'object', 'additionalProperties': f_schema}, definitions
-        else:
-            # The dict values are Any, no need to declare it
-            return {'type': 'object'}, definitions
+        if key_pattern:
+            # Dict keys have a regex pattern
+            # f_schema might be a schema or empty dict, add it either way
+            dict_schema.update({'patternProperties': {key_pattern: f_schema}})
```

Comment:
since we're only setting on item use setitem `dict_schema['patternProperties']` here and below.